lib/puppet/provider/elasticsearch_plugin/plugin.rb

```diff
@@ -0,0 +1,157 @@
+$LOAD_PATH.unshift(File.join(File.dirname(__FILE__),"..","..",".."))
+
+Puppet::Type.type(:elasticsearch_plugin).provide(:plugin) do
+  desc "A provider for the resource type `elasticsearch_plugin`,
+        which handles plugin installation"
+
+  commands :plugin => '/usr/share/elasticsearch/bin/plugin'
+  commands :es => '/usr/share/elasticsearch/bin/elasticsearch'
+
+  def exists?
+    es_version
+    if !File.exists?(pluginfile)
+      debug "Plugin file #{pluginfile} does not exist"
+      return false
+    elsif File.exists?(pluginfile) && readpluginfile != pluginfile_content
+      debug "Got #{readpluginfile} Expected #{pluginfile_content}"
+      self.destroy
+      false
+    else
+      debug "Plugin exists"
+      return true
+    end
+  end
+
+  def pluginfile_content
+    if is2x?
+      items = @resource[:name].split("/")
+      if items.count == 1 # Official plugin
+        version = plugin_version(@resource[:name])
+        return "#{@resource[:name]}/#{version}"
+      else
+        return @resource[:name]
+      end
+    else
+      return @resource[:name]
+    end
+  end
+
+  def pluginfile
+    File.join(@resource[:plugin_dir], plugin_name(@resource[:name]), '.name')
+  end
+
+  def writepluginfile
+    File.open(pluginfile, 'w') do |file|
+      file.write pluginfile_content
+    end
+  end
+
+  def readpluginfile
+    f = File.open(pluginfile)
+    f.readline
+  end
+
+  def install1x
+    if !@resource[:url].nil?
+      commands = [ plugin_name(@resource[:name]), '--url', @resource[:url] ]
+    elsif !@resource[:source].nil?
+      commands = [ plugin_name(@resource[:name]), '--url', "file://#{@resource[:source]}" ]
+    else
+      commands = [ @resource[:name] ]
+    end
+    commands
+  end
+
+  def install2x
+    if !@resource[:url].nil?
+      commands = [ @resource[:url] ]
+    elsif !@resource[:source].nil?
+      commands = [ "file://#{@resource[:source]}" ]
+    else
+      commands = [ @resource[:name] ]
+    end
+    commands
+  end
+
+  def create
+    es_version
+    commands = []
+    commands << @resource[:proxy_args] if @resource[:proxy_args]
+    commands << 'install'
+    commands << install1x if is1x?
+    commands << install2x if is2x?
+    
+    plugin(commands)
+    writepluginfile
+  end
+
+  def destroy
+    plugin(['remove', @resource[:name]])
+  end
+
+  def es_version
+    return @es_version if @es_version
+    begin
+      version = es('-v') # ES 1.x
+    rescue
+      version = es('--version') # ES 2.x
+    rescue
+      raise "Unknown ES version. Got #{version.inspect}"
+    ensure
+      @es_version = version.scan(/\d+\.\d+\.\d+(?:\-\S+)?/).first
+      debug "Found ES version #{@es_version}"
+    end
+  end
+
+  def is1x?
+    Puppet::Util::Package.versioncmp(@es_version, '2.0.0') < 0
+  end
+
+  def is2x?
```

Comment:
`(Puppet::Util::Package.versioncmp(@es_version, '2.0.0') >= 0) and
(Puppet::Util::Package.versioncmp(@es_version, '3.0.0') < 0)` would
probably work to limit `is2x?`, then do similar for `is3x?` when it's out
and use that were possible. The question that I can't answer is would it be
better to then have to update other functions to say `if (x.is2x? or
x.is3x?)` or add additional if/else branches; or to rename the function and
change those instances now to `if x.isnot1x?`. Those may still need updated
after 3x is released anyway.

Or, since tons of code may need touched then anyway, fix `is2x?` to limit
`2.0.0<=@es_version<3.0.0` and worry about it later :)


Rob Nelson
rnelson0@gmail.com

On Mon, Dec 7, 2015 at 11:45 AM, Richard Pijnenburg <
notifications@github.com> wrote:

> In lib/puppet/provider/elasticsearch_plugin/plugin.rb
> <https://github.com/elastic/puppet-elasticsearch/pull/507#discussion_r46843953>
> :
>
> > +      version = es('-v') # ES 1.x
> > +    rescue
> > +      version = es('--version') # ES 2.x
> > +    rescue
> > +      raise "Unknown ES version. Got #{version.inspect}"
> > +    ensure
> > +      @es_version = version.scan(/\d+\.\d+\.\d+(?:\-\S+)?/).first
> > +      debug "Found ES version #{@es_version}"
> > +    end
> > +  end
> > +
> > +  def is1x?
> > +    Puppet::Util::Package.versioncmp(@es_version, '2.0.0') < 0
> > +  end
> > +
> > +  def is2x?
>
> It could happen that 3.x might have other weird things requiring a is3x?
> Not sure how to solve that in a nice way.
> Any idea's?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/puppet-elasticsearch/pull/507/files#r46843953>
> .
>